desci-server/src/controllers/nodes/publish.ts

```diff
@@ -98,49 +102,270 @@ export const publish = async (req: PublishRequest, res: Response<PublishResBody>
-        commitId: commitId ?? '',
-        uuid: ensureUuidEndsWithDot(uuid),
-        status: PublishTaskQueueStatus.WAITING,
-      },
-    });
-
     return res.send({
       ok: true,
-      taskId: publishTask.id,
+      taskId: publishTask?.id,
     });
   } catch (err) {
     logger.error({ err }, '[publish::publish] node-publish-err');
     return res.status(400).send({ ok: false, error: err.message });
-  }
+  };
+};
+
+/**
+ * Synchronously perform publish steps before returning.
+ *
+ * This is generally fast because the ceramic update is already done,
+ * but two cases are a bit slower:
+ * 1. first time publish (wait on backend tx)
+ * 2. dpid upgrade (wait on backend tx)
+ *
+ * Semantically, these can both be made fire-and-forget promises if we can
+ * manage without instantly having the dPID alias available in this function.
+*/
+const syncPublish = async (
+  ceramicStream: string,
+  commitId: string,
+  node: Node,
+  owner: User,
+  cid: string,
+  uuid: string,
+  manifest: ResearchObjectV1,
+): Promise<void> => {
+  const logger = parentLogger.child({
+    module: 'NODE::syncPublish',
+    uuid,
+    cid,
+    ceramicStream,
+    commitId,
+  });
+
+  const latestNodeVersion = await prisma.nodeVersion.findFirst({
+    where: {
+      id: -1,
+      nodeId: node.id,
+    },
+    orderBy: {
+      id: "desc",
+    },
+  });
+
+  // Prevent duplicating the NodeVersion entry if the latest version is the same as the one we're trying to publish, as a draft save is triggered before publishing
+  const latestNodeVersionId = latestNodeVersion?.manifestUrl === cid
+    ? latestNodeVersion.id
+    : -1;
+
+  const nodeVersion = await prisma.nodeVersion.upsert({
+    where: {
+      id: latestNodeVersionId,
+    },
+    update: {
+      commitId,
+    },
+    create: {
+      nodeId: node.id,
+      manifestUrl: cid,
+      commitId,
+    },
+  });
+
+  // first time we see a stream for this node, make sure we bind it in the db
+  if (!node.ceramicStream) {
+    logger.trace(`[publish:publish] setting streamID ${ceramicStream} on node ${uuid}`);
+    await setCeramicStream(uuid, ceramicStream);
+  } else if (node.ceramicStream !== ceramicStream) {
+    logger.warn(
+      // This is unexpected and weird, but important to know if it occurs
+      `[publish:publish] stream on record does not match passed streamID`,
+      { database: node.ceramicStream, ceramicStream },
+    );
+  };
+
+  const legacyDpid = manifest.dpid?.id ? parseInt(manifest.dpid.id) : undefined;
+  let dpidAlias: number = node.dpidAlias;
+
+  // Do dataRef and dPID registration operations concurrently
+  const promises = [];
+
+  if (!dpidAlias) {
+    // The only reason this isn't just fire-and-forget is that we want the dpid
+    // for the discord notification, which won't be available otherwise for
+    // first time publishes.
+    promises.push(
+      createOrUpgradeDpidAlias(legacyDpid, ceramicStream, uuid)
+      .then(dpid => dpidAlias = dpid)
+    );
+  };
```

Comment:
Here the `dpidAlias` value is used to figure out if we need to create a new/upgrade a legacy dPID